File: Python/1047-student_list_for_course.py


if __name__ == '__main__':
    n_student, n_course = [int(i) for i in input().split()]
    course_to_student = [[] for _ in range(n_course + 1)]
    for _ in range(n_student):
        name, _, *reg_courses = input().split()
        for course in reg_courses:
            course_to_student[int(course)].append(name)
    for course in range(1, n_course + 1):
        reg_students = course_to_student[course]
        print(course, len(reg_students))
        for student in sorted(reg_students):
            print(student)

File: C++/1047-student_list_for_course.cc
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>


int main() {

    std::ios::sync_with_stdio(false);

    int n_student;
    std::size_t n_course;
    std::cin >> n_student >> n_course;

    std::vector<std::vector<std::string>> course_to_stu(n_course + 1);
    while (n_student--) {
        std::string name;
        int n_reg;
        std::cin >> name >> n_reg;
        while (n_reg--) {
            std::size_t course;
            std::cin >> course;
            course_to_stu[course].push_back(name);
        }
    }

    for (std::size_t course=1; course<=n_course; ++course) {
        std::vector<std::string> students = course_to_stu[course];
        std::cout << course << ' ' << students.size() << '\n';
        std::sort(students.begin(), students.end());
        for (std::string &student : students) {
            std::cout << student << '\n';
        }
    }

    return 0;
}
